src/base/asyncfilestorage.cpp

```diff
@@ -0,0 +1,113 @@
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "asyncfilestorage.h"
+
+#include <QDebug>
+#include <QMetaObject>
+#include <QMutexLocker>
+#include <QSaveFile>
+#include <QTimer>
+
+AsyncFileStorage::AsyncFileStorage(const QDir &storageDir, QFile *lockFile, QObject *parent)
+    : QObject(parent)
+    , m_storageDir(storageDir)
+    , m_lockFile(lockFile)
+{
+}
+
+AsyncFileStorage::~AsyncFileStorage()
+{
+    m_lockFile->close();
+    m_lockFile->remove();
+    delete m_lockFile;
+}
+
+AsyncFileStorage *AsyncFileStorage::create(const QString &storageFolder, QObject *parent)
+{
+    QDir storageDir(storageFolder);
+    QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
+    if (storageDir.exists() || storageDir.mkpath(storageDir.absolutePath())) {
+        if (lockFile->open(QFile::WriteOnly))
+            return new AsyncFileStorage(storageDir, lockFile, parent);
+    }
+
+    delete lockFile;
+    return nullptr;
+}
+
+void AsyncFileStorage::store(const QString &fileName, const QByteArray &data)
+{
+    QMutexLocker locker(&m_jobsMutex);
+    m_jobs.insert(fileName, data);
+    qDebug() << "AsyncFileStorage: Added job for" << fileName;
+    if (m_isIdle) {
+        m_isIdle = false;
+        // The function will be run in the thread of receiver (this AsyncFileStorage object).
+        QTimer::singleShot(0, this, &AsyncFileStorage::processPendingJobs);
+    }
+}
+
+void AsyncFileStorage::processPendingJobs()
+{
+    m_jobsMutex.lock();
+    QStringList fileNames = m_jobs.keys();
+    m_jobsMutex.unlock();
+
+    while (!fileNames.isEmpty()) {
+        foreach (const QString &fileName, fileNames) {
+            m_jobsMutex.lock();
+            const QByteArray data = m_jobs.take(fileName);
```

Comment:
Strong guarantee. 
Perhaps the most simple in itself. If we leave the job in the queue in case of failure, we simply overwrite it as you normally would, if updated data become available.

User interaction. 
We can't implement this for non-interactive component. We can only delegate it in the interactive layer, but how? Now we do not have a suitable notifications subsystem so that it will require to spoil the interfaces of the affected classes by ugly inappropriate signals/methods.

One more question.
Let's say we have the ability to communicate with the user. What can he do in this case? As I said, it's either some kind of temporary glitch that does not require any response, except, perhaps, fix this event in the log and attempts to repeat writing some time ago (agree, I have to implement it), or there is a serious malfunction of the whole system (what to do in this case?).

In General, if you don't have a ready solution (not abstract advice), let's leave this for a separate topic. It doesn't break current application behavior.